src/upgrades/mod.rs

```diff
@@ -129,94 +260,175 @@ pub trait UpgradeStore {
+    let current = upgrade.to();
+    let data_dir = &config.data_dir;
+    let upgrade_dir = config.upgrade_data_dir();
+
+    // cas -> arch-cas-{old-version}
+    // upgrade-data/cas -> cas
+    // upgrade-data/ca_objects -> ca_objects
+    // set cas/version
+
+    let cas = data_dir.join(CASERVER_DIR);
+    let cas_version = cas.join("version");
+    let cas_arch = data_dir.join(format!("arch-{}-{}", CASERVER_DIR, from));
+    let cas_upg = upgrade_dir.join(CASERVER_DIR);
+    let ca_objects = data_dir.join(CA_OBJECTS_DIR);
+    let ca_objects_upg = upgrade_dir.join(CA_OBJECTS_DIR);
+
+    move_dir_if_exists(&cas, &cas_arch)?;
+    move_dir_if_exists(&cas_upg, &cas)?;
+    move_dir_if_exists(&ca_objects_upg, &ca_objects)?;
+    file::save_json(&current, &cas_version)?;
+
+    // pubd -> arch-pubd-{old-version}
+    // upgrade-data/pubd -> pubd
+    // upgrade-data/pubd_objects -> pubd_objects
+    // set pubd/version
+
+    let pubd = data_dir.join(PUBSERVER_DIR);
+    let pubd_version = pubd.join("version");
+    let pubd_arch = data_dir.join(format!("arch-{}-{}", PUBSERVER_DIR, from));
+    let pubd_upg = upgrade_dir.join(PUBSERVER_DIR);
+    let pubd_objects = data_dir.join(PUBSERVER_CONTENT_DIR);
+    let pubd_objects_upg = upgrade_dir.join(PUBSERVER_CONTENT_DIR);
+
+    move_dir_if_exists(&pubd, &pubd_arch)?;
+    move_dir_if_exists(&pubd_upg, &pubd)?;
+    move_dir_if_exists(&pubd_objects_upg, &pubd_objects)?;
+    file::save_json(&current, &pubd_version)?;
+
+    // done, clean out the migration dir
+    file::remove_dir_all(&upgrade_dir)?;
+
+    // move the dirs
+    fn move_dir_if_exists(from: &Path, to: &Path) -> UpgradeResult<()> {
+        if from.exists() {
+            std::fs::rename(from, to).map_err(|e| {
+                UpgradeError::Custom(format!(
+                    "Could not rename directory from: {} to: {}. Error: {}",
+                    from.to_string_lossy(),
+                    to.to_string_lossy(),
+                    e
+                ))
+            })
+        } else {
+            Ok(())
+        }
     }
-}
 
-/// Should be called when Krill starts, before the KrillServer is initiated
-pub fn pre_start_upgrade(config: Arc<Config>) -> Result<(), UpgradeError> {
-    upgrade_data_to_0_9_0(config)
+    Ok(())
 }
 
-/// Should be called when the KrillServer is initiated, before the webserver is started
+/// Should be called when the KrillServer is initiated, before the web server is started
 /// and operators can make changes.
-pub async fn post_start_upgrade(config: &Config, server: &KrillServer) -> Result<(), UpgradeError> {
-    if needs_upgrade(&config.data_dir, "cas", KrillVersion::candidate(0, 9, 3, 2)) {
+pub async fn post_start_upgrade(upgrade_versions: &UpgradeVersions, server: &KrillServer) -> Result<(), UpgradeError> {
+    if upgrade_versions.from() < &KrillVersion::candidate(0, 9, 3, 2) {
         info!("Reissue ROAs on upgrade to force short EE certificate subjects in the objects");
```

Comment:
Wasn't there some idea about not doing this?

Reply:
Issue #772. It's conditional - will skip for instances with many CAs. After this is merged